Ensure that query expressions fully bind expressions

333fred opened the pull request:
We weren't calling CheckValue on the result of binding `from ... in x` (specifically on the `x` component). This means we were missing a couple of things that CheckValue does:
* Realize property accesses as getters, and ensure that a getter actually exists.
* Ensure default arguments are bound for these property accesses.

Fixes https://github.com/dotnet/roslyn/issues/50316.

333fred commented:
@dotnet/roslyn-compiler for review.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/QueryTests.cs:
```diff
@@ -4134,15 +4134,15 @@ static void Main()
+                // file.cs(15,45): error CS0119: 'ParentNamespace.ConsoleApp' is a namespace, which is not valid in the given context
                 //                 var z = /*<bind>*/from c in NSAlias select 3/*</bind>*/;
-                Diagnostic(ErrorCode.ERR_BadSKunknown, "NSAlias").WithArguments("NSAlias", "namespace").WithLocation(15, 45)
+                Diagnostic(ErrorCode.ERR_BadSKunknown, "NSAlias").WithArguments("ParentNamespace.ConsoleApp", "namespace").WithLocation(15, 45)
```
Note that we're now following the alias to the underlying namespace. This matches our behavior in the rest of the compiler.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/QueryTests.cs:
```diff
@@ -4442,5 +4442,93 @@ class Program
         }
+
+        [Fact, WorkItem(50316, "https://github.com/dotnet/roslyn/issues/50316")]
+        public void SetOnlyProperty()
```
This has been crashing for who knows how long. I guess no one has attempted to query syntax a set-only property before and bothered to tell us that it crashed the compiler for them.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/QueryTests.cs:
```diff
@@ -4442,5 +4442,93 @@ class Program
+}
+", options: TestOptions.ReleaseExe);
+
+            comp.VerifyEmitDiagnostics();
```
>comp.VerifyEmitDiagnostics(); [](start = 12, length = 29)

I think we want to execute the code and observe expected behavior at runtime.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/QueryTests.cs:
```diff
@@ -4442,5 +4442,93 @@ class Program
+           select @int;
+", references: new[] { vb.EmitToImageReference() }, options: TestOptions.ReleaseExe);
+
+            comp.VerifyEmitDiagnostics();
```
>comp.VerifyEmitDiagnostics(); [](start = 12, length = 29)

Here as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Query.cs:
```diff
@@ -39,7 +39,7 @@ internal BoundExpression BindQuery(QueryExpressionSyntax node, DiagnosticBag dia
 
             QueryTranslationState state = new QueryTranslationState();
-            state.fromExpression = MakeMemberAccessValue(boundFromExpression, diagnostics);
+            state.fromExpression = CheckValue(MakeMemberAccessValue(boundFromExpression, diagnostics), BindValueKind.RValueOrType, diagnostics);
```
>CheckValue [](start = 35, length = 10)

Addition of this call here feels premature and inconsistent with how we deal with binding of a regular member access. Specifically, in case of ```TypeOrValueExpression```:
```
                    case BoundKind.TypeOrValueExpression:
                        {
                            // CheckValue call will occur in ReplaceTypeOrValueReceiver.
```  
It is quite possible that we are missing to call CheckValue at some point, but it feels like that should happen later in the pipeline. Addition of a new BindValueKind just to accommodate this call site feels suspicious and, for me, is one of the indicators of possibly doing work at wrong time/place.

333fred replied:
I'm not really sure where "later in the pipeline" would be here though. The next thing we do with `state.fromExpression` is potentially wrap it with a call to `Cast`, and it needs to have had default arguments/rvalueness checked by that point. In examining other uses of CheckValue, we call it once we know what context we're going to be using the value in, and this feels like the right to me. As to adding a new kind, most places in C# don't let you use a literal type as a receiver, but query syntax does, so it feels a bit special in that regard.

---
In reply to: [554076229](https://github.com/dotnet/roslyn/pull/50322#discussion_r554076229) [](ancestors = 554076229)

AlekseyTs replied:
> I'm not really sure where "later in the pipeline" would be here though. The next thing we do with state.fromExpression is potentially wrap it with a call to Cast, and it needs to have had default arguments/rvalueness checked by that point. In examining other uses of CheckValue, we call it once we know what context we're going to be using the value in, and this feels like the right to me. As to adding a new kind, most places in C# don't let you use a literal type as a receiver, but query syntax does, so it feels a bit special in that regard.

Query expressions in C# is a syntactic translation. The thing from ```from``` is a receiver, the ```Cast``` or any other query operator is a member access on that receiver. As pretty much anywhere in the language, a receiver of a member access is allowed to be a type, it doesn't feel like there is anything special about queries in that regard. My intuition is that process of binding that synthetic member access should pretty much do what ```BindMemberAccessWithBoundLeft``` does, or, perhaps, call it eventually. I am assuming that, if the query is manually translated into method calls, binding succeeds for that translated form, but fails for the untranslated form. For the problematic receiver, I would try to understand where binding for a query diverges from binding for a regular member access in terms of checking for default arguments/rvalueness for the receiver. That might give us some ideas.


---
In reply to: [554201536](https://github.com/dotnet/roslyn/pull/50322#discussion_r554201536) [](ancestors = 554201536,554076229)

333fred replied:
I've moved the checks into `MakeQueryInvocation`, where the rest of the related receiver checks occur. In particular, I discovered that we allow querying over type parameters, which should be invalid. An email to compat council has tentative support, but I'll hold off on merging until that's confirmed and a test insertion passes.

---
In reply to: [554217774](https://github.com/dotnet/roslyn/pull/50322#discussion_r554217774) [](ancestors = 554217774,554201536,554076229)

AlekseyTs commented:
Done with review pass (commit 1)

333fred commented:
@AlekseyTs addressed feedback, thanks.

---
In reply to: [756882776](https://github.com/dotnet/roslyn/pull/50322#issuecomment-756882776) [](ancestors = 756882776)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Query.cs:
```diff
@@ -784,6 +791,14 @@ protected BoundCall MakeQueryInvocation(CSharpSyntaxNode node, BoundExpression r
             }
+            else
+            {
+                var checkedUltimateReceiver = CheckValue(ultimateReceiver, BindValueKind.RValue, diagnostics);
```
>CheckValue [](start = 46, length = 10)

It feels like this could be premature for a BoundKind.TypeOrValueExpression receiver. Need to make sure we have tests for a "Color Color" case, covering both ways of resolving the receiver.

AlekseyTs commented:
>             while (ultimateReceiver.Kind == BoundKind.QueryClause) ultimateReceiver = ((BoundQueryClause)ultimateReceiver).Value;

Consider changing this to the following in order to improve readability:
```
while (ultimateReceiver.Kind == BoundKind.QueryClause)
{
    ultimateReceiver = ((BoundQueryClause)ultimateReceiver).Value;
}
```

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Query.cs:728 in 09047b2. [](commit_id = 09047b213d233daff7644dfc31d5d16624eb26e0, deletion_comment = False)

AlekseyTs commented:
Done with review pass (commit 3)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [759520627](https://github.com/dotnet/roslyn/pull/50322#issuecomment-759520627) [](ancestors = 759520627)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Query.cs:
```diff
@@ -797,6 +820,24 @@ protected BoundCall MakeQueryInvocation(CSharpSyntaxNode node, BoundExpression r
+            {
+                if (receiver is BoundQueryClause query)
+                {
+                    return query.Update(
```
>return query.Update( [](start = 20, length = 20)

Is this code path covered by an y test?

333fred replied:
Yes. Both indexed property cases hit this: they receiver is wrapped in a query clause unconditionally on line 64 in this file.

---
In reply to: [556915303](https://github.com/dotnet/roslyn/pull/50322#discussion_r556915303) [](ancestors = 556915303)

AlekseyTs reviewed (approved):
LGTM (commit 5), assuming CI is passing

333fred commented:
@dotnet/roslyn-compiler for a second review please.

333fred commented:
@dotnet/roslyn-compiler for a second review please.

333fred commented:
CloudBuild passed: https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/297749.
